stelaval/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,176 @@
+package assignment2;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class BinaryTree<T> {
+
+  private Node<T> root;
+  private Set<Node<T>> nodeSet;
+
+  public BinaryTree() {
+    nodeSet = new HashSet<>();
+  }
+
+  /**
+   * The nodes of the binary tree are stored in a set in order to avoid duplicate values.
+   * 
+   * @param newNode is the node to be added to the set of nodes
+   */
+  public void add(Node<T> newNode) {
+    if (root == null) {
+      root = newNode;
+    }
+
+    if (nodeSet.contains(newNode)) {
+      throw new IllegalArgumentException();
+    } else {
+      nodeSet.add(newNode);
+    }
+  }
+
+  /**
+   * Prints all the ancestors of a node
+   * 
+   * @param key is a node which ancestors are printed
+   * 
+   * @throws Exception if given key is not present in a tree
+   */
+  public void printAncestors(Node<T> key) throws Exception {
+    if (nodeSet.contains(key)) {

```

Comment:
What conditions would have to be met to avoid having such checks? It is good to do sanity checks like that in code, but it is even better to design class in a way that would make them impossible to happen.